File: backend/semgrep_rules/tests/javascript/kisa_sw_01_sql_injection.js
async function unsafeConcat(db, req) {
  const userId =
    req.query.id;

  // ruleid: kisa.sw01.javascript.sql-client-string-built-query
  return db.query("SELECT * FROM users WHERE id = " + userId);
}


async function unsafeTemplate(db, req) {
  const username =
    req.body.username;

  // ruleid: kisa.sw01.javascript.sql-client-string-built-query
  return db.query(`SELECT * FROM users WHERE username = '${username}'`);
}


async function unsafeIntermediate(db, req) {
  const status =
    req.params.status;

  const sql =
    "SELECT * FROM users WHERE status = '" + status + "'";

  const copiedSql =
    sql;

  // ruleid: kisa.sw01.javascript.sql-client-string-built-query
  return db.execute(copiedSql);
}


async function unsafeDirectSql(db, req) {
  const sql =
    req.body.sql;

  // ruleid: kisa.sw01.javascript.sql-client-string-built-query
  return db.query(sql);
}


async function unsafeRaw(db, req) {
  const orderBy =
    req.query.sort;

  const sql =
    `SELECT * FROM users ORDER BY ${orderBy}`;

  // ruleid: kisa.sw01.javascript.sql-client-string-built-query
  return db.raw(sql);
}


async function safeMysqlBinding(db, req) {
  const userId =
    req.query.id;

  // ok: kisa.sw01.javascript.sql-client-string-built-query
  return db.query(
    "SELECT * FROM users WHERE id = ?",
    [userId]
  );
}


async function safePgBinding(db, req) {
  const userId =
    req.query.id;

  // ok: kisa.sw01.javascript.sql-client-string-built-query
  return db.query(
    "SELECT * FROM users WHERE id = $1",
    [userId]
  );
}


async function safeHardcoded(db) {
  // ok: kisa.sw01.javascript.sql-client-string-built-query
  return db.query("SELECT id, name FROM users");
}


async function safeStructuredOrm(User, req) {
  const userId =
    req.query.id;

  // ok: kisa.sw01.javascript.sql-client-string-built-query
  return User.findOne({
    where: {
      id: userId,
    },
  });
}

File: backend/semgrep_rules/tests/python/kisa_sw_01_sql_injection.py
def unsafe_concat(cursor, request):
    user_id = request.GET.get("id")

    # ruleid: kisa.sw01.python.dbapi-string-built-query
    return cursor.execute("SELECT * FROM users WHERE id = " + user_id)


def unsafe_percent(cursor, request):
    username = request.POST.get("username")

    sql = (
        "SELECT * FROM users WHERE username = '%s'"
        % username
    )

    # ruleid: kisa.sw01.python.dbapi-string-built-query
    return cursor.execute(sql)


def unsafe_format(cursor, request):
    status = request.args.get("status")

    sql = "SELECT * FROM users WHERE status = '{}'".format(
        status
    )

    # ruleid: kisa.sw01.python.dbapi-string-built-query
    return cursor.execute(sql)


def unsafe_fstring(cursor, request):
    role = request.form.get("role")

    sql = f"SELECT * FROM users WHERE role = '{role}'"

    # ruleid: kisa.sw01.python.dbapi-string-built-query
    return cursor.execute(sql)


def unsafe_multi_step(cursor, request):
    order_by = request.GET.get("sort")

    sql = (
        "SELECT * FROM users ORDER BY "
        + order_by
    )

    copied_sql = sql
    second_copy = copied_sql

    # ruleid: kisa.sw01.python.dbapi-string-built-query
    return cursor.execute(second_copy)


def unsafe_direct_sql(cursor, request):
    sql = request.POST.get("sql")

    # ruleid: kisa.sw01.python.dbapi-string-built-query
    return cursor.execute(sql)


def unsafe_django_raw(User, request):
    name = request.GET.get("name")

    sql = (
        "SELECT * FROM auth_user WHERE username = '"
        + name
        + "'"
    )

    # ruleid: kisa.sw01.python.dbapi-string-built-query
    return User.objects.raw(sql)


def safe_parameter_binding(cursor, request):
    user_id = request.GET.get("id")

    # ok: kisa.sw01.python.dbapi-string-built-query
    return cursor.execute(
        "SELECT * FROM users WHERE id = %s",
        [user_id],
    )


def safe_named_binding(cursor, request):
    username = request.GET.get("username")

    # ok: kisa.sw01.python.dbapi-string-built-query
    return cursor.execute(
        "SELECT * FROM users WHERE username = %(username)s",
        {
            "username": username,
        },
    )


def safe_hardcoded(cursor):
    # ok: kisa.sw01.python.dbapi-string-built-query
    return cursor.execute(
        "SELECT id, username FROM users"
    )


def safe_django_orm(User, request):
    user_id = request.GET.get("id")

    # ok: kisa.sw01.python.dbapi-string-built-query
    return User.objects.filter(
        id=user_id
    )
